### PR TITLE
Transaction Semantic Validation

### PR DESCRIPTION
**Pull request by SeanCheatham**
## Purpose
- Tetra Transaction Inputs include a Box Value and a Proposition that must match the data associated with the Transaction Output that it spends
- Transaction Inputs must reference Transaction Outputs that have not yet been spent along the current chain
## Approach
- Added ContextlessValidationAlgebra and ContextualValidationAlgebra
  - The future plan is to re-use these algebras for Header Validation as well
- Added TransactionSemanticValidationAlgebra and TransactionSemanticValidation interpreter which leverages BoxState
- Renamed SyntacticValidation to TransactionSyntacticValidation
## Testing
- New TransactionSemanticValidation tests
## Tickets
- Closes _#2078_

### CODE REVIEW

**Review by scasplte2 — APPROVED** [review]
🔬

File: algebras/src/main/scala/co/topl/algebras/ContextlessValidation.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package co.topl.algebras
+
+import cats.data.ValidatedNec
+
+trait ContextlessValidation[F[_], E, T] {
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply algebras/src/main/scala/co/topl/algebras/ContextlessValidation.scala]: Good catch - fixed

File: ledger/src/main/scala/co/topl/ledger/algebras/TransactionSemanticValidationAlgebra.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package co.topl.ledger.algebras
+
+import cats.data.ValidatedNec
+import co.topl.algebras.ContextualValidationAlgebra
+import co.topl.models.{Box, Transaction, TypedIdentifier}
+
+trait TransactionSemanticValidationAlgebra[F[_]]
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply ledger/src/main/scala/co/topl/ledger/algebras/TransactionSemanticValidationAlgebra.scala:7]: By scope-tracking, do you mean with the type arguments?

File: ledger/src/main/scala/co/topl/ledger/algebras/TransactionSemanticValidationAlgebra.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    extends ContextualValidationAlgebra[F, InvalidSemanticError, Transaction, TypedIdentifier] {
+
+  /**
+   * Determines the semantic integrity of the given transaction within the context of the given parentBlockId
+   */
+  def validateSemantics(parentBlockId: TypedIdentifier)(
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply ledger/src/main/scala/co/topl/ledger/algebras/TransactionSemanticValidationAlgebra.scala]: Fixed (using `validate` for consistency)

File: ledger/src/main/scala/co/topl/ledger/interpreters/TransactionSemanticValidation.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                  .toNestedValidatedNec
+                  .value
+                  .map(_.leftMap(_.flatten).as(transaction))
+              case (invalid, _) => invalid.pure[F]
+            }
+      }
+    )
+
+  /**
+   * Does the box value included on the input of _this_ transaction match the value on the output of the spent transaction?
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply ledger/src/main/scala/co/topl/ledger/interpreters/TransactionSemanticValidation.scala:44]: Updated the comment to be a bit more in line with your suggestion

File: ledger/src/main/scala/co/topl/ledger/interpreters/TransactionSemanticValidation.scala [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+   *
+   * Does the proposition's evidence included on the input of _this_ transaction match the evidence associated
+   * with the Spending Address which owns the box being spent?
+   */
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply ledger/src/main/scala/co/topl/ledger/interpreters/TransactionSemanticValidation.scala:48]: Agreed - tough to strike a balance between specificity and readability

File: ledger/src/main/scala/co/topl/ledger/interpreters/TransactionSemanticValidation.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      .map(spentTransaction =>
+        // Did the output referenced by this input ever exist?  (Not a spend-ability check, just existence)
+        spentTransaction.outputs
+          .get(input.boxId.transactionOutputIndex)
+          .toValidNec[InvalidSemanticError](InvalidSemanticErrors.UnspendableBox(input.boxId))
+          .ensure(NonEmptyChain(InvalidSemanticErrors.InputDataMismatch(input)))(spentOutput =>
+            // Does the value claimed in the input match the value defined on the referenced output?
+            spentOutput.value === input.value &&
+            // Does the proposition claimed in the input contain the same evidence that is defined on the
+            // referenced output's address?
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply ledger/src/main/scala/co/topl/ledger/interpreters/TransactionSemanticValidation.scala]: I updated to `spent output` and tried to use that terminology consistently.  It's probably the most accurate terminology, though perhaps not the most understandable?

**Pull request merged**